components/supervisor/pkg/supervisor/supervisor.go

```diff
@@ -646,6 +671,83 @@ func startAPIEndpoint(ctx context.Context, cfg *Config, wg *sync.WaitGroup, serv
 	l.Close()
 }
 
+func tunnelOverWebSocket(tunneled *ports.TunneledPortsService, conn *gitpod.WsConn) {
+	hostKey, err := generateHostKey()
+	if err != nil {
+		log.WithError(err).Error("tunnel: failed to generate host key")
+		conn.Close()
+		return
+	}
+	config := &ssh.ServerConfig{
+		NoClientAuth: true,
+	}
+	config.AddHostKey(hostKey)
+	sshConn, chans, reqs, err := ssh.NewServerConn(conn, config)
+	if err != nil {
+		log.WithError(err).Error("tunnel: ssh connection handshake failed")
+		return
+	}
+	go func() {
+		conn.Wait()
+		sshConn.Close()
+	}()
+	go ssh.DiscardRequests(reqs)
+	go func() {
+		for ch := range chans {
+			go tunnelOverSSH(conn.Ctx, tunneled, ch)
+		}
+	}()
+	err = sshConn.Wait()
+	if err != nil {
+		log.WithError(err).Error("tunnel: ssh connection failed")
+	}
+}
+
+func generateHostKey() (ssh.Signer, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+	return ssh.NewSignerFromKey(key)
+}
+
+func tunnelOverSSH(ctx context.Context, tunneled *ports.TunneledPortsService, newCh ssh.NewChannel) {
+	tunnelReq := &api.TunnelPortRequest{}
+	err := proto.Unmarshal(newCh.ExtraData(), tunnelReq)
+	if err != nil {
+		log.WithError(err).Error("tunnel: invalid ssh chan request")
+		newCh.Reject(ssh.Prohibited, err.Error())
+		return
+	}
+
+	tunnel, err := tunneled.EstablishTunnel(ctx, tunnelReq.ClientId, tunnelReq.Port, tunnelReq.TargetPort)
+	if err != nil {
+		log.WithError(err).Error("tunnel: failed to establish")
+		newCh.Reject(ssh.Prohibited, err.Error())
+		return
+	}
+	defer tunnel.Close()
+
+	sshChan, reqs, err := newCh.Accept()
+	if err != nil {
+		log.WithError(err).Error("tunnel: accepting ssh channel failed")
+		return
+	}
+	defer sshChan.Close()
+	go ssh.DiscardRequests(reqs)
+	done := make(chan struct{})
+	var once sync.Once
+	go func() {
+		io.Copy(sshChan, tunnel)
```

Comment:
Might be useful to log errors from the Copy

Reply:
I tried first, but then there were always only noise, like using closed network on disconnections. I checked what other similar libraries are doing:
- https://github.com/jpillora/chisel/blob/92d90be68a989377daf61294ef7458612d10da8c/share/cio/pipe.go#L19-L24
- https://github.com/gliderlabs/ssh/blob/30ec06db4e743ac9f827a69c8b8cfb84064a6dc7/tcpip.go#L59-L64

Reply:
then, use `_ = io.Copy` to indicate the explicit choice that the error is ignored. Linting should have complained here, too :/